### ISSUE TITLE
how to do large scale multi-task multi-label classification?

### ISSUE THREAD

**Issue by yil8** [fchollet/keras#1331]
Hi, I'm wondering if there are any handy ways to do large scale multi-task multi-label classification using multi-layer-perceptron?

My X has ~2000 features and my Y has ~1000 targets(tasks). Each target takes 3 possible labels(-1, 0, 1). I have looked into the graph model. It seems I need to manually type:

graph.add_output(name='output1') 
graph.add_output(name='output2')
....
graph.add_output(name='output1000')

~1000 times. And to specify softmax as loss, I may also need to type something like 

graph.compile(optimizer='sgd', loss={'output1':'softmax', 'output2':'softmax', ..., 'output1000':'softmax'}). Note that, each softmax output itself corresponding to a Nx3 matrix in my three-label case.

This is fine for small number of targets(<10?). But I'm wondering if there are cleaner ways to do it with 1000 targets?

Thanks so much!

**Comment by fchollet** [comment]
> This is fine for small number of targets(<10?). But I'm wondering if there are cleaner ways to do it with 1000 targets?

You could do the reverse --3 tasks with 1000 labels each, and use binary_crossentropy for each task (since the label probabilities are independent). 

Alternatively, use a single task with a 3000-dimensional output and write your own loss function that does softmax by groups of 3 vectors.

Even simpler: use a single task with a 3000-dimensional output and use binary_crossentropy as loss.

**Comment by yil8** [comment]
That's a great idea. Thanks!

**Comment by alyato** [comment]
@yil8 Do you share your code? Or Do you give me example how to implement  your idear by graph model. Thanks.

**Comment by yil8** [comment]
@alyato I actually gave up doing multi-task multi-label classification and ended up doing multi-task 1-vs-all binary classification instead.

**Comment by alyato** [comment]
@yil8 , I find some guys also use the keras to do . [https://github.com/fchollet/keras/issues/741](url)
And there are some questions i can't understand.
1.how to load the multi-label dataset.  
2.Do you give me one simple example to run
Thanks.

**Comment by yil8** [comment]
@alyato the issue link is broken. 

So assuming you are doing 3 labels classification.  Basically, what you need to do is a transformation from

          X             y
[1, 2, 3]             1
[4, 5, 6]            2
[7, 8, 9]            3
......

to 


       X             y
[1, 2, 3]          [1, 0, 0]
[4, 5, 6]         [0, 1, 0]
[7, 8, 9]         [0, 0, 1]
......

Then you do multi-task binary classification

**Comment by alyato** [comment]
@yil8  [https://github.com/fchollet/keras/issues/741]
My dataset is that one image have multi-label.
y has three label [0,1,2], one image can has labels[3,1],or [1,2]
`x`            `y`
`[1,2,3]`                  `3,1`
`[4,5,6]`                  `1,2`
Then I need  tranform them to 
`x`            `y`
`[1,2,3]`      `[1,0,1]`
`[4,5,6]`      `[1,1,0]`
Is it right? Thanks

**Comment by yil8** [comment]
Yeah, I think this works. You basically have one bit for each label indicating whether it's on or off

**Comment by CreativePao** [comment]
@alyato Hi, I'm glad to talk to you. I see you're doing multi-label classification. Now I also doing this subject, It's just that I'm dealing with text. I want to ask you, did you succeed? Can you tell me what layers and parameters you used the neural network? if you agree, could you show me your code？Thx!

**Comment by alyato** [comment]
@CreativePao Im not doing multi-label now and using the code by suraj-deshmukh. 
The link is  https://suraj-deshmukh.github.io/Multi-Label-Image-Classification/. You can check it. Thanks.
